File: P3213/Popov_368679/lab5/methods/Lagrange.js

export class Lagrange {
  static solve(xCur, data) {
    const size = data.size;
    let answer = 0;
    for (let i = 0; i < size; i++) {
      let polynomial = 1;
      for (let j = 0; j < size; j++) {
        if (i !== j) {
          polynomial *=
            (xCur - data.points[j].x) / (data.points[i].x - data.points[j].x);
        }
      }
      answer += data.points[i].y * polynomial;
    }
    return answer;
  }
}

File: P3208/Dashkevich_367187/lab5/src/methods/lagrange.py
from abstract_polynom import Polynomial
from utils5 import *


class Lagrange_Polynomial(Polynomial):
    def __init__(self, points):
        super().__init__(points, list(zeros(len(points))), "lagrange")

        for i in range(len(points)):
            numerator = []
            buff = points[i][1]

            for j in range(len(points)):
                if i == j:
                    continue
                buff /= points[i][0] - points[j][0] # считаем знаменатель
                numerator.append(-1 * points[j][0]) # собираем числитель

            _polynom = expand_brackets(numerator)
            _polynom = [elem * buff for elem in _polynom]
            self.koofs = [self.koofs[i] + _polynom[i] for i in range(len(self.koofs))]